pkg/image/trigger/buildconfigs/buildconfigs.go

```diff
@@ -0,0 +1,194 @@
+package buildconfigs
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+
+	"github.com/golang/glog"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	triggerapi "github.com/openshift/origin/pkg/image/api/v1/trigger"
+	"github.com/openshift/origin/pkg/image/trigger"
+)
+
+// calculateBuildConfigTriggers transforms a build config into a set of image change triggers.
+// It uses synthetic field paths since we don't need to generically transform the config.
+func calculateBuildConfigTriggers(bc *buildapi.BuildConfig) []triggerapi.ObjectFieldTrigger {
+	var triggers []triggerapi.ObjectFieldTrigger
+	for _, t := range bc.Spec.Triggers {
+		if t.ImageChange == nil {
+			continue
+		}
+		var (
+			fieldPath string
+			from      *kapi.ObjectReference
+		)
+		if t.ImageChange.From != nil {
+			from = t.ImageChange.From
+			fieldPath = "spec.triggers"
+		} else {
+			from = buildutil.GetInputReference(bc.Spec.Strategy)
+			fieldPath = "spec.strategy.*.from"
+		}
+		if from == nil || from.Kind != "ImageStreamTag" || len(from.Name) == 0 {
+			continue
+		}
+
+		// add a trigger
+		triggers = append(triggers, triggerapi.ObjectFieldTrigger{
+			From: triggerapi.ObjectReference{
+				Name:       from.Name,
+				Namespace:  from.Namespace,
+				Kind:       from.Kind,
+				APIVersion: from.APIVersion,
+			},
+			FieldPath: fieldPath,
+		})
+	}
+	return triggers
```

Comment:
Origin objects (bc, dc) have triggers fields and we use those, explicitly (for backwards compatibility, as well). Annotations are only used where we don't have those fields, iow. mostly k8s objects. That's at least my understanding.